File: src/main/java/com/lessmarkup/interfaces/recordmodel/RecordModel.java
package com.lessmarkup.interfaces.recordmodel;

import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.interfaces.data.DataObject;

public abstract class RecordModel<T extends RecordModel> {
    private final Class<? extends ModelCollection<T>> collectionType;
    private final Class<? extends DataObject> dataType;
    private final String titleTextId;
    private final boolean submitWithCaptcha;
    private long id;
    
    protected RecordModel(String titleTextId, Class<? extends ModelCollection<T>> collectionType, Class<? extends DataObject> dataType, boolean submitWithCaptcha) {
        this.collectionType = collectionType;
        this.dataType = dataType;
        this.titleTextId = titleTextId;
        this.submitWithCaptcha = submitWithCaptcha;
    }
    
    protected RecordModel(String titleTextId, Class<? extends ModelCollection<T>> collectionType, Class<? extends DataObject> dataType) {
        this(titleTextId, collectionType, dataType, false);
    }
    
    protected RecordModel(String titleTextId, Class<? extends ModelCollection<T>> collectionType) {
        this(titleTextId, collectionType, null);
    }
    
    protected RecordModel(String titleTextId) {
        this(titleTextId, null);
    }

    protected RecordModel(String titleTextId, boolean submitWithCaptcha) {
        this(titleTextId, null, null, submitWithCaptcha);
    }

    protected RecordModel(Class<? extends ModelCollection<T>> collectionType, Class<? extends DataObject> dataType) {
        this(null, collectionType, dataType);
    }
    
    protected RecordModel() {
        this(null);
    }
    
    public Class<? extends DataObject> getDataType() {
        return dataType;
    }
    
    public String getTitleTextId() {
        return titleTextId;
    }
    
    public boolean getSubmitWithCaptcha() {
        return submitWithCaptcha;
    }
    
    public ModelCollection<T> createCollection() {
        return DependencyResolver.resolve(this.collectionType);
    }
    
    public long getId() {
        return this.id;
    }
    
    public void setId(long id) {
        this.id = id;
    }
}

File: src/main/scala/com/lessmarkup/interfaces/recordmodel/RecordModel.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.interfaces.recordmodel

import com.lessmarkup.framework.helpers.DependencyResolver
import com.lessmarkup.interfaces.data.DataObject

abstract class RecordModel[T](titleTextId: String, collectionType: Class[_ <: ModelCollection[T]], dataType: Class[_ <: DataObject], submitWithCaptcha: Boolean) {
  var id: Long = 0L

  protected def this(titleTextId: String, collectionType: Class[_ <: ModelCollection[T]], dataType: Class[_ <: DataObject]) {
    this(titleTextId, collectionType, dataType, false)
  }

  protected def this(titleTextId: String, collectionType: Class[_ <: ModelCollection[T]]) {
    this(titleTextId, collectionType, null)
  }

  protected def this(titleTextId: String) {
    this(titleTextId, null)
  }

  protected def this(titleTextId: String, submitWithCaptcha: Boolean) {
    this(titleTextId, null, null, submitWithCaptcha)
  }

  protected def this(collectionType: Class[_ <: ModelCollection[T]], dataType: Class[_ <: DataObject]) {
    this(null, collectionType, dataType)
  }

  protected def this() {
    this(null)
  }

  def getDataType: Class[_ <: DataObject] = dataType
  def getTitleTextId: String = titleTextId
  def getSubmitWithCaptcha: Boolean = submitWithCaptcha
  def createCollection: ModelCollection[T] = DependencyResolver.resolve(this.collectionType)
}
